pkg/bgpv1/manager/reconcilerv2/crd_status.go

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package reconcilerv2
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8s_types "k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/ptr"
+
+	daemon_k8s "github.com/cilium/cilium/daemon/k8s"
+	"github.com/cilium/cilium/pkg/bgpv1/agent/mode"
+	"github.com/cilium/cilium/pkg/bgpv1/manager/instance"
+	"github.com/cilium/cilium/pkg/bgpv1/types"
+	"github.com/cilium/cilium/pkg/k8s"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	k8s_client "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/k8s/resource"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/time"
+)
+
+const (
+	CRDStatusUpdateInterval = 5 * time.Second
+)
+
+type StatusReconciler struct {
+	lock.Mutex
+
+	Logger            logrus.FieldLogger
+	ClientSet         k8s_client.Clientset
+	LocalNodeResource daemon_k8s.LocalCiliumNodeResource
+
+	nodeName      string
+	desiredStatus *v2alpha1.CiliumBGPNodeStatus
+	runningStatus *v2alpha1.CiliumBGPNodeStatus
+}
+
+type StatusReconcilerIn struct {
+	cell.In
+
+	Job       job.Group
+	ClientSet k8s_client.Clientset
+	Logger    logrus.FieldLogger
+	LocalNode daemon_k8s.LocalCiliumNodeResource
+}
+
+type StatusReconcilerOut struct {
+	cell.Out
+
+	Reconciler StateReconciler `group:"bgp-state-reconciler"`
+}
+
+func NewStatusReconciler(in StatusReconcilerIn) StatusReconcilerOut {
+	// CRD Status reconciler is disabled if there is no kubernetes support
+	if !in.ClientSet.IsEnabled() {
+		return StatusReconcilerOut{}
+	}
+
+	r := &StatusReconciler{
+		Logger:            in.Logger.WithField(types.ReconcilerLogField, "CRD_Status"),
+		LocalNodeResource: in.LocalNode,
+		ClientSet:         in.ClientSet,
+	}
+
+	in.Job.Add(job.OneShot("bgp-crd-status-initialize", func(ctx context.Context, health cell.Health) error {
+		r.Logger.Debug("Initializing")
+
+		for event := range r.LocalNodeResource.Events(ctx) {
+			switch event.Kind {
+			case resource.Upsert:
+				r.Lock()
+				r.nodeName = event.Object.GetName()
+				r.Unlock()
+			}
+			event.Done(nil)
+		}
+		return nil
+	}))
+
+	in.Job.Add(job.OneShot("bgp-crd-status-update-job", func(ctx context.Context, health cell.Health) (err error) {
+		r.Logger.Debug("Update job running")
+
+		ticker := time.NewTicker(CRDStatusUpdateInterval)
+		defer ticker.Stop()
```

Comment:
The purpose of rate-limiting the state reconciliation is to protect the API server. Then, we should do exponential backoff on failure and make a random jitter for every interval to avoid synchronized access. We should also make this interval configurable. Otherwise, users have nothing to do when they notice that BGP CPlane is overwhelming their API server.

Reply:
Added exponential backoff on failure. I'd like to discuss more about configurable interval for updating CRD status. Should it be configurable via CRD or via cilium configmap ?

Reply:
I vote for the configMap option either. I agree that this kind of the "admin" option doesn't fit into the ClusterConfig. I think this interval is about how much delay users can accept for status report and I believe this value rarely changes.

On the other hand, this is also a trade off between API server load and propagation delay. So, users may want to change this value when their API server is overloaded. However, because changing the value requires BGP session reset (without GR), the operational cost is significant, so we should try to not overload the API server as much as possible even if users set a short interval.

I think it's just about following the best practices like:
- Add jitter to the interval to avoid synchronized write from multiple nodes
- ✅ Exponentially backoff when the API call fails
- ✅ Avoid API call whenever possible (e.g. no diff)

Reply:
Added this [commit](https://github.com/cilium/cilium/pull/33411/commits/ced206e433c10eb5e1f2e20f63fe4b9ef67a8477) to introduce ticker with jitter. Using external dependency for tickerWithJitter implementation.